### PR TITLE
Adjust the length of blob cache docs for Lucene metadata files

### PR DESCRIPTION
**Pull request by tlrx**
Today searchable snapshots `IndexInput` implementations use the _blob store cache_ to cache the first `4096` bytes of every Lucene files. After some experiments we think that we could adjust the length of the cached data depending of the Lucene file that is read, caching up to 64KB for Lucene _metadata_ files (ie files that are fully read when a Directory is opened) and only 1KB for other files.

The files that are cached up to 64KB are the files with the following extensions:

>         "cfe", // compound file's entry table
>         "dvm", // doc values metadata file
>         "fdm", // stored fields metadata file
>         "fnm", // field names metadata file
>         "kdm", // Lucene 8.6 point format metadata file
>         "nvm", // norms metadata file
>         "tmd", // Lucene 8.6 terms metadata file
>         "tvm", // terms vectors metadata file
>         "vem"  // Lucene 9.0 indexed vectors metadata

The 64KB limit can be configured on a per index basis  through a new index setting.

This change is extracted from #69283 and does not address the caching of CFS files.

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java [hunk 1]
```diff
@@ -67,6 +77,35 @@
...
+        builder.put(CacheService.SNAPSHOT_CACHE_RANGE_SIZE_SETTING.getKey(), blobCacheMaxLength);
+        builder.put(CacheService.SNAPSHOT_CACHE_RECOVERY_RANGE_SIZE_SETTING.getKey(), blobCacheMaxLength);
+
+        // Frozen (shared cache) cache should be large enough to not cause direct reads
+        builder.put(SnapshotsService.SNAPSHOT_CACHE_SIZE_SETTING.getKey(), ByteSizeValue.ofMb(128));
+        // Align ranges to match the blob cache max length
+        builder.put(SnapshotsService.SNAPSHOT_CACHE_REGION_SIZE_SETTING.getKey(), blobCacheMaxLength);
+        builder.put(SnapshotsService.SHARED_CACHE_RANGE_SIZE_SETTING.getKey(), blobCacheMaxLength);
+        builder.put(FrozenCacheService.FROZEN_CACHE_RECOVERY_RANGE_SIZE_SETTING.getKey(), blobCacheMaxLength);
+        cacheSettings = builder.build();
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java:100]: I'm trying to randomize these settings.

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java [hunk 2]
```diff
@@ -79,28 +118,25 @@ protected int numberOfReplicas() {
...
 
+        final int numberOfDocs = scaledRandomIntBetween(0, 20_000);
+        if (numberOfDocs > 0) {
+            final List<IndexRequestBuilder> indexRequestBuilders = new ArrayList<>();
+            for (int i = numberOfDocs; i > 0; i--) {
+                XContentBuilder builder = XContentFactory.smileBuilder();
+                builder.startObject().field("text", randomRealisticUnicodeOfCodepointLengthBetween(5, 50)).field("num", i).endObject();
+                indexRequestBuilders.add(client().prepareIndex(indexName).setSource(builder));
+            }
+            indexRandom(true, true, true, indexRequestBuilders);
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java:138]: It now uses `dummyDocs` in order to generate deletions

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/blobstore/cache/BlobStoreCacheService.java [hunk 3]
```diff
@@ -155,4 +159,46 @@ public void onFailure(Exception e) {
...
+     * @param maxMetadataLength the maximum accepted length for metadata files
+     *
+     * @return the header {@link ByteRange}
+     */
+    public static ByteRange computeBlobCacheByteRange(String fileName, long fileLength, ByteSizeValue maxMetadataLength) {
+        final String fileExtension = IndexFileNames.getExtension(fileName);
+        if (METADATA_FILES_EXTENSIONS.contains(fileExtension)) {
+            final long maxAllowedLengthInBytes = maxMetadataLength.getBytes();
+            if (fileLength > maxAllowedLengthInBytes) {
+                logger.warn(
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/blobstore/cache/BlobStoreCacheService.java:193]: Maybe too verbose?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java [hunk 4]
```diff
@@ -185,6 +187,32 @@
...
+                    "failed to parse value [%s] for setting [%s], must be a multiple of [%s] bytes",
+                    value.getStringRep(),
+                    SNAPSHOT_BLOB_CACHE_METADATA_FILES_MAX_LENGTH,
+                    BufferedIndexInput.BUFFER_SIZE
+                );
+                throw new IllegalArgumentException(message);
+            }
+        },
+        Setting.Property.IndexScope,
+        Setting.Property.NodeScope,
```
Comment (ywelsch) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:213]: Let's make it purely indexScope?

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java [hunk 5]
```diff
@@ -67,6 +77,35 @@
...
+        final Settings.Builder builder = Settings.builder();
+        // Cold (full copy) cache should be unlimited to not cause evictions
+        builder.put(CacheService.SNAPSHOT_CACHE_SIZE_SETTING.getKey(), new ByteSizeValue(Long.MAX_VALUE, ByteSizeUnit.BYTES));
+        // Align ranges to match the blob cache max length
+        builder.put(CacheService.SNAPSHOT_CACHE_RANGE_SIZE_SETTING.getKey(), blobCacheMaxLength);
+        builder.put(CacheService.SNAPSHOT_CACHE_RECOVERY_RANGE_SIZE_SETTING.getKey(), blobCacheMaxLength);
+
+        // Frozen (shared cache) cache should be large enough to not cause direct reads
+        builder.put(SnapshotsService.SNAPSHOT_CACHE_SIZE_SETTING.getKey(), ByteSizeValue.ofMb(128));
+        // Align ranges to match the blob cache max length
```
Comment (ywelsch) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java:96]: why have this alignment?

**Review by ywelsch — COMMENTED** [review]
Thanks for opening this PR. I've done a first pass and left some comments.

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java [hunk 6]
```diff
@@ -163,106 +209,98 @@ public void testBlobStoreCache() throws Exception {
...
             SearchableSnapshotsStatsAction.INSTANCE,
             new SearchableSnapshotsStatsRequest()
         ).actionGet().getStats()) {
             for (final SearchableSnapshotShardStats.CacheIndexInputStats indexInputStats : shardStats.getStats()) {
                 // we read the header of each file contained within the .cfs file, which could be anywhere
-                final boolean mayReadMoreThanHeader = indexInputStats.getFileExt().equals("cfs");
-                if (indexInputStats.getTotalSize() <= BlobStoreCacheService.DEFAULT_CACHED_BLOB_SIZE * 2
-                    || mayReadMoreThanHeader == false) {
+                final boolean mayReadMoreThanHeader = indexInputStats.getFileExt().equals("cfs")
+                    || indexInputStats.getFileExt().equals("cfe");
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java]: It's a leftover from debugging sessions, I removed them in 1d9cda0d11f175f803b517a8032092548b1ba6d3

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/blobstore/cache/BlobStoreCacheService.java [hunk 7]
```diff
@@ -155,4 +159,46 @@ public void onFailure(Exception e) {
             listener.onFailure(e);
         }
     }
+
+    private static final Set<String> METADATA_FILES_EXTENSIONS = Set.of(
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/blobstore/cache/BlobStoreCacheService.java]: Sure, I pushed 40f7557a42a0123ad1e322241279b9ce61e5e768

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java [hunk 8]
```diff
@@ -162,25 +163,18 @@ protected void doReadInternal(ByteBuffer b) throws IOException {
 
             // We try to use the cache index if:
             // - the file is small enough to be fully cached
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java]: Addressed in 0cc9e9d428cd94926e786e0605ad697174eeff2e

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java [hunk 9]
```diff
@@ -274,11 +268,12 @@ protected void doReadInternal(ByteBuffer b) throws IOException {
                 final Future<Integer> readFuture = cacheFile.readIfAvailableOrPending(indexCacheMiss, channel -> {
                     final int indexCacheMissLength = toIntBytes(indexCacheMiss.length());
 
+                    // TODO: should we use BigArrays?
                     // We assume that we only cache small portions of blobs so that we do not need to:
                     // - use a BigArrays for allocation
                     // - use an intermediate copy buffer to read the file in sensibly-sized chunks
                     // - release the buffer once the indexing operation is complete
-                    assert indexCacheMissLength <= COPY_BUFFER_SIZE : indexCacheMiss;
+                    // assert indexCacheMissLength <= COPY_BUFFER_SIZE : indexCacheMiss;
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java]: Sure -> 6e70ff3a0c057975add184a6ce654b5e3687e38f

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java [hunk 10]
```diff
@@ -274,11 +268,12 @@ protected void doReadInternal(ByteBuffer b) throws IOException {
                 final Future<Integer> readFuture = cacheFile.readIfAvailableOrPending(indexCacheMiss, channel -> {
                     final int indexCacheMissLength = toIntBytes(indexCacheMiss.length());
 
+                    // TODO: should we use BigArrays?
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java]: Ok - 6e70ff3a0c057975add184a6ce654b5e3687e38f

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java [hunk 11]
```diff
@@ -600,7 +595,8 @@ public IndexInput slice(String sliceDescription, long offset, long length) {
             length,
             cacheFileReference,
             defaultRangeSize,
-            recoveryRangeSize
+            recoveryRangeSize,
+            ByteRange.EMPTY // TODO implement blob cache for CFS
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java]: Yes, I reworded this TODO in 5d28df834b81a0823abda9abf60944647ad8c195 but it should be adressed in a follow up.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java (hunk 4 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java]: Good catch, I pushed b5f2463ba4c80ddf4d87d969e969aced2d70f849

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/blobstore/cache/BlobStoreCacheService.java [hunk 12]
```diff
@@ -41,7 +45,7 @@
 
     private static final Logger logger = LogManager.getLogger(BlobStoreCacheService.class);
 
-    public static final int DEFAULT_CACHED_BLOB_SIZE = ByteSizeUnit.KB.toIntBytes(4);
+    public static final int DEFAULT_CACHED_BLOB_SIZE = ByteSizeUnit.KB.toIntBytes(1);
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/blobstore/cache/BlobStoreCacheService.java:53]: Sorry for the delayed response, I had to deal with other duties.

Looking at the current code in 7.x, reducing the size of cached blob might throw an exception in Cached/FrozenIndexInput at this line:

> final BytesRefIterator cachedBytesIterator = cachedBlob.bytes().slice(toIntBytes(position), length).iterator();

in the case `position` is larger than one or two buffered reads.

I suggest 
- to keep the 4KB/8KB limit for non metadata files as long as there is at least a node in version < 7.13 in the cluster
- to write some BWC test that ensure indices are correctly assigned during rolling upgrades

**Comment by tlrx** [comment]
Thanks @ywelsch for your review. I updated the code according to your comments. 

Concerning https://github.com/elastic/elasticsearch/pull/69431#discussion_r581009695 and the behavior with previous versions, I followed my suggestion in https://github.com/elastic/elasticsearch/pull/69431#discussion_r582646610 and I added a rolling upgrade test named `SearchableSnapshotsRollingUpgradeIT`.

This integration test is composed of 4 tests: 
- two are tests that mount a snapshot as an index in the previous version and verifies that the index is recovered and searchable on mixed and upgraded version.
- the two other tests are a bit tricky and try to test the behavior of the blob cache when cached blob documents are created from an old version or a new version. Those tests only runs in mixed version cluster. I validated that it throws the exception I suspected in https://github.com/elastic/elasticsearch/pull/69431#discussion_r582646610.

This is ready for another review.

**Review by ywelsch — APPROVED** [review]
LGTM. Thanks Tanguy!

**Comment by tlrx** [comment]
Thanks Yannick! 

I'm going to backport this down to 7.12 and adjusts the BWC version OLD_CACHED_BLOB_SIZE_VERSION in follow up PRs.

**Pull request merged**